### ISSUE TITLE
`JavaVisitor#maybeAddImport()` doesn't check for conflicts

### ISSUE THREAD

**Issue by knutwannheden** [openrewrite/rewrite#2580]
Many recipes use `JavaVisitor#maybeAddImport()` which adds an import for a type if that type (if it then also is referenced). But it appears that the `AddImport` recipe doesn't check for conflicts.

Assuming we have a source with an import for `either `org.assertj.core.util.Objects`, `org.testng.collections.Objects`, or `com.google.common.base.Objects`, then a recipe like `ReplaceLambdaWithMethodReference` cannot really add an import for `java.util.Objects` like it currently does. In this specific case not only should the `maybeAddImport()` call be a no-op, but also the method references to `Objects::isNull` and `Objects::nonNull` should be fully qualified.

Maybe the safest course of action would be for recipes to use fully-qualified references in any Java templates and then have these get shortened in the presence of matching imports and potential imports (registered by the recipe using `maybeAddImport()`).

**Comment by knutwannheden** [comment]
A similar situation may arise when the modified source is in a package which already contains a type with the same name as the type being imported. Adding the import would be wrong as it would invalidate any existing unqualified references to the package type.

**Comment by jkschneider** [comment]
In many cases I imagine text-based JavaTemplate will be replaced by `JavaTemplate.compile()` which resembles the Refaster style templates (see https://github.com/openrewrite/feat-rewrite-templating/blob/2a48e3901da036f9fe857f4f8481aacbd8549383/src/test/java/org/openrewrite/java/template/AutoTemplateTest.java#L47-L49). Any references in the template will be unambiguous.

**Comment by knutwannheden** [comment]
Yes, in Refaster-style templates all references will of course be unambiguous, which is a big advantage. I will take a look at that repo. Thanks for the link!

I imagine there could still be use cases for the current textual Java templates, so it might still be worth solving this issue. Also, I think the `AddImport` recipe should somehow detect and deal with these aliasing situations.

**Comment by rpau** [comment]
@knutwannheden do you remember in what recipe context this bug occurs?

**Comment by knutwannheden** [comment]
@rpau Yes, it is in the `ReplaceLambdaWithMethodReference` recipe mentioned in the description.

Going forward I think it would be very nice if we could make `JavaTemplate` smarter. What do I mean by this? I think it would make sense if the template string would typically use fully qualified names for all classes and static members and then there would be style options to control what the code produced for the template then looks like and what kind of imports get added (static imports vs. qualified references). `JavaTemplate` could then do the conflict resolution and decide when a fully qualified reference is required or not. With fully qualified references in the template string the recipe author would then not have to use `JavaVisitor#maybeAddImport()` nor `JavaTemplate.Builder#imports()`, which I think would be great.

(Rather than doing the import conflict resolution as part of `JavaTemplate` we could also apply a separate visitor to shorten all qualified references which don't result in conflicts (by adding corresponding imports) after a source file has been modified using a template. This seems like the better approach to me.)
